v-next/hardhat/test/internal/builtin-plugins/solidity/build-system/compiler/downloader.ts

```diff
@@ -0,0 +1,446 @@
+    describe("getCompiler", function () {
+      it("should throw when trying to get a compiler that doesn't exist in the compiler list", async function () {
+        await downloader.updateCompilerListIfNeeded(new Set([]));
+
+        await assertRejectsWithHardhatError(
+          () => downloader.getCompiler("0.4.12"),
+          HardhatError.ERRORS.INTERNAL.ASSERTION_ERROR,
+          {
+            message: "Trying to get a compiler 0.4.12 before it was downloaded",
+          },
+        );
+      });
+
+      it("should throw when trying to get a compiler that's in the compiler list but hasn't been downloaded yet", async function () {
+        await downloader.updateCompilerListIfNeeded(
+          new Set(["0.4.12", "0.4.13"]),
+        );
+
+        await downloader.downloadCompiler("0.4.12");
+
+        await assertRejectsWithHardhatError(
+          () => downloader.getCompiler("0.4.13"),
+          HardhatError.ERRORS.INTERNAL.ASSERTION_ERROR,
+          {
+            message: "Trying to get a compiler 0.4.13 before it was downloaded",
+          },
+        );
+      });
+
+      it("should throw if the native compiler can't be verified", async function () {
+        let hasDownloadedOnce = false;
+        const platform = CompilerDownloader.getCompilerPlatform();
+        const mockDownloader = new CompilerDownloader(
+          platform,
+          process.cwd(),
+          async (_url, destination, _requestOptions, _dispatcherOptions) => {
+            if (!hasDownloadedOnce) {
+              hasDownloadedOnce = true;
+              const longVersion = "0.4.12+mock";
+              const compilersList = {
+                builds: [
+                  {
+                    path:
+                      platform === CompilerPlatform.WINDOWS
+                        ? "solc.exe"
+                        : "solc",
+                    version: "0.4.12",
+                    longVersion,
+                    build: "build",
+                    keccak256:
+                      "0x87c2d362de99f75a4f2755cdaaad2d11bf6cc65dc71356593c445535ff28f43d",
+                    urls: [],
+                    platform,
+                  },
+                ],
+                releases: {
+                  "0.4.12": "0.4.12+mock",
+                },
+                latestRelease: "0.4.12",
+              };
+
+              await fs.ensureDir(path.dirname(destination));
+              await fs.writeJsonFile(destination, compilersList);
+              return;
+            }
+
+            await fs.ensureDir(path.dirname(destination));
+            await fs.writeUtf8File(destination, "asd");
+          },
+        );
+
+        await mockDownloader.updateCompilerListIfNeeded(new Set(["0.4.12"]));
+
+        assertRejectsWithHardhatError(
+          () => mockDownloader.downloadCompiler("0.4.12"),
+          HardhatError.ERRORS.SOLIDITY.INVALID_DOWNLOAD,
+          {
+            remoteVersion: "0.4.12+commit.194ff033",
+          },
+        );
```

Comment:
I've seen this check being flaky already as we sometimes get `HardhatError.ERRORS.SOLIDITY.DOWNLOAD_FAILED` instead. What do you think about adding a retry mechanism to the downloader?

Reply:
I had another look at that and I don't quite understand why this is being flaky. We're using a mock downloader in this test case, which just writes to the disk. I assumed it was flaky because of the network use, but we're not even touching it here.

I'm trying to capture the full error we get when this flakes on another branch. I'll report back here when I find something.

I still think a retry mechanism might be an idea worth exploring, but I would definitely like to understand the source of flakiness in this test case better.

Reply:
Found it! It was throwing the dir doesn't exist error because we weren't awaiting the assert 🤦 We should be able to catch it with a linting rule. 

```suggestion
        await assertRejectsWithHardhatError(
          () => mockDownloader.downloadCompiler("0.4.12"),
          HardhatError.ERRORS.SOLIDITY.INVALID_DOWNLOAD,
          {
            remoteVersion: "0.4.12+commit.194ff033",
          },
        );
```